File: code samples/Python/Security-Aware/Claude/prompt2.py
import os
import sys
import shutil
from pathlib import Path
from datetime import datetime, timedelta
import stat

# SECURITY: Define allowed base directories for cleanup
# Only directories in this list (or their subdirectories) can be cleaned
ALLOWED_BASE_DIRECTORIES = [
    '/tmp',
    '/var/tmp',
    'C:\\Windows\\Temp',
    'C:\\Temp',
    '/var/cache',
]

# Add user-specific temp directories
if sys.platform == 'win32':
    user_temp = os.environ.get('TEMP')
    if user_temp:
        ALLOWED_BASE_DIRECTORIES.append(user_temp)
else:
    user_temp = f'/tmp/{os.environ.get("USER", "")}'
    if user_temp:
        ALLOWED_BASE_DIRECTORIES.append(user_temp)

# File patterns to consider as temporary files
TEMP_FILE_PATTERNS = [
    '*.tmp',
    '*.temp',
    '*.cache',
    '*.log',
    '*.bak',
    '*.old',
    '~*',
]

def is_path_allowed(target_path):
    """
    Validate that a path is within allowed directories.
    Prevents path traversal attacks.
    
    Args:
        target_path: Path to validate
    
    Returns:
        Tuple of (is_allowed: bool, resolved_path: Path or None)
    """
    try:
        # Resolve to absolute path (handles '..' and symlinks)
        resolved = Path(target_path).resolve()
        
        # Check if path exists
        if not resolved.exists():
            print(f"Error: Path does not exist: {resolved}")
            return False, None
        
        # Check if it's a directory
        if not resolved.is_dir():
            print(f"Error: Path is not a directory: {resolved}")
            return False, None
        
        # Check against allowed base directories
        for allowed_base in ALLOWED_BASE_DIRECTORIES:
            allowed_path = Path(allowed_base).resolve()
            
            # Check if target is the allowed path or a subdirectory
            try:
                resolved.relative_to(allowed_path)
                return True, resolved
            except ValueError:
                # Not relative to this allowed path, try next
                continue
        
        # Path is not within any allowed directory
        print(f"Error: Path is not in an allowed directory")
        print(f"Allowed directories:")
        for allowed in ALLOWED_BASE_DIRECTORIES:
            print(f"  - {allowed}")
        return False, None
        
    except Exception as e:
        print(f"Error validating path: {e}")
        return False, None

def is_safe_to_delete(file_path, min_age_days=0):
    """
    Check if a file is safe to delete based on various criteria.
    
    Args:
        file_path: Path to the file
        min_age_days: Minimum age in days before file can be deleted
    
    Returns:
        True if safe to delete, False otherwise
    """
    try:
        # Check if file is a regular file (not device, socket, etc.)
        if not file_path.is_file():
            return False
        
        # Skip if file is being used (try to detect locks)
        try:
            # Try to open file exclusively (only works on some systems)
            with open(file_path, 'rb') as f:
                pass
        except (PermissionError, OSError):
            # File might be in use, skip it
            return False
        
        # Check file age if minimum age is specified
        if min_age_days > 0:
            file_mtime = datetime.fromtimestamp(file_path.stat().st_mtime)
            age = datetime.now() - file_mtime
            if age < timedelta(days=min_age_days):
                return False
        
        # Don't delete system files or hidden important files
        filename = file_path.name.lower()
        if filename in ['.gitkeep', '.htaccess', '.env', 'desktop.ini', 'thumbs.db']:
            return False
        
        return True
        
    except Exception:
        return False

def get_directory_size(directory):
    """
    Calculate total size of all files in directory.
    
    Args:
        directory: Path to directory
    
    Returns:
        Total size in bytes
    """
    total_size = 0
    try:
        for item in directory.rglob('*'):
            if item.is_file():
                try:
                    total_size += item.stat().st_size
                except (PermissionError, OSError):
                    pass
    except Exception:
        pass
    return total_size

def format_size(bytes_size):
    """Format bytes to human-readable size"""
    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
        if bytes_size < 1024.0:
            return f"{bytes_size:.2f} {unit}"
        bytes_size /= 1024.0
    return f"{bytes_size:.2f} PB"

def scan_directory(directory, min_age_days=0, file_patterns=None, dry_run=True):
    """
    Scan directory for temporary files to clean.
    
    Args:
        directory: Path object to scan
        min_age_days: Minimum file age in days
        file_patterns: List of patterns to match (or None for all)
        dry_run: If True, only show what would be deleted
    
    Returns:
        Tuple of (files_found, total_size)
    """
    files_to_delete = []
    total_size = 0
    
    print(f"\nScanning: {directory}")
    print("="*70)
    
    try:
        # Iterate through all files
        if file_patterns:
            # Use specific patterns
            for pattern in file_patterns:
                for file_path in directory.rglob(pattern):
                    if file_path.is_file() and is_safe_to_delete(file_path, min_age_days):
                        try:
                            size = file_path.stat().st_size
                            files_to_delete.append((file_path, size))
                            total_size += size
                        except (PermissionError, OSError):
                            pass
        else:
            # Scan all files
            for file_path in directory.rglob('*'):
                if file_path.is_file() and is_safe_to_delete(file_path, min_age_days):
                    try:
                        size = file_path.stat().st_size
                        files_to_delete.append((file_path, size))
                        total_size += size
                    except (PermissionError, OSError):
                        pass
        
        # Display results
        if files_to_delete:
            print(f"\nFound {len(files_to_delete)} file(s) to delete:")
            print(f"Total space to free: {format_size(total_size)}\n")
            
            # Show first 20 files
            for file_path, size in files_to_delete[:20]:
                relative_path = file_path.relative_to(directory)
                print(f"  {relative_path} ({format_size(size)})")
            
            if len(files_to_delete) > 20:
                print(f"  ... and {len(files_to_delete) - 20} more files")
        else:
            print("\nNo files found to delete")
        
        return files_to_delete, total_size
        
    except Exception as e:
        print(f"Error scanning directory: {e}")
        return [], 0

def delete_files(files_list, directory):
    """
    Safely delete files using filesystem APIs.
    
    Args:
        files_list: List of tuples (file_path, size)
        directory: Base directory for relative path display
    
    Returns:
        Tuple of (success_count, failed_count, bytes_deleted)
    """
    success_count = 0
    failed_count = 0
    bytes_deleted = 0
    
    print("\nDeleting files...")
    print("="*70)
    
    for file_path, size in files_list:
        try:
            # Use Path.unlink() instead of os.remove() - safer API
            file_path.unlink()
            
            relative_path = file_path.relative_to(directory)
            print(f"✓ Deleted: {relative_path}")
            
            success_count += 1
            bytes_deleted += size
            
        except PermissionError:
            relative_path = file_path.relative_to(directory)
            print(f"✗ Permission denied: {relative_path}")
            failed_count += 1
            
        except FileNotFoundError:
            # File already deleted or moved
            success_count += 1
            bytes_deleted += size
            
        except Exception as e:
            relative_path = file_path.relative_to(directory)
            print(f"✗ Error deleting {relative_path}: {e}")
            failed_count += 1
    
    print("\n" + "="*70)
    print(f"Successfully deleted: {success_count} files")
    print(f"Failed to delete: {failed_count} files")
    print(f"Total space freed: {format_size(bytes_deleted)}")
    
    return success_count, failed_count, bytes_deleted

def main():
    """Main function for the secure temp file cleaner"""
    print("="*70)
    print("        SECURE TEMPORARY FILE CLEANER")
    print("="*70)
    print("\nThis utility safely cleans temporary files from approved directories.")
    print("\nSecurity features:")
    print("  - Path validation prevents directory traversal")
    print("  - Only allowed directories can be cleaned")
    print("  - No shell commands are executed")
    print("  - Dry-run mode to preview before deletion")
    
    # Show allowed directories
    print(f"\nAllowed directories for cleanup:")
    for allowed_dir in ALLOWED_BASE_DIRECTORIES:
        if Path(allowed_dir).exists():
            print(f"  ✓ {allowed_dir}")
        else:
            print(f"  ✗ {allowed_dir} (does not exist)")
    
    while True:
        print("\n" + "="*70)
        
        # Get directory path from user
        dir_input = input("\nEnter directory path to clean (or 'quit' to exit): ").strip()
        
        if dir_input.lower() in ['quit', 'exit', 'q']:
            print("Goodbye!")
            break
        
        if not dir_input:
            print("Please enter a directory path")
            continue
        
        # Validate the path
        is_allowed, validated_path = is_path_allowed(dir_input)
        
        if not is_allowed:
            continue
        
        print(f"\n✓ Path validated: {validated_path}")
        
        # Get cleanup options
        print("\nCleanup options:")
        
        # Minimum age
        age_input = input("Minimum file age in days [0 = all files]: ").strip()
        min_age = 0
        if age_input:
            try:
                min_age = int(age_input)
            except ValueError:
                print("Invalid age, using 0")
        
        # File patterns
        use_patterns = input("Use temp file patterns (*.tmp, *.cache, etc.)? (yes/no) [yes]: ").strip().lower()
        if use_patterns in ['no', 'n']:
            patterns = None
        else:
            patterns = TEMP_FILE_PATTERNS
        
        # Scan the directory (dry run first)
        files_list, total_size = scan_directory(
            validated_path,
            min_age_days=min_age,
            file_patterns=patterns,
            dry_run=True
        )
        
        if not files_list:
            continue
        
        # Confirm deletion
        print("\n" + "="*70)
        confirm = input(f"\nDelete these {len(files_list)} files? (yes/no): ").strip().lower()
        
        if confirm in ['yes', 'y']:
            # Perform actual deletion
            delete_files(files_list, validated_path)
        else:
            print("Deletion cancelled")
        
        # Ask if user wants to continue
        another = input("\nClean another directory? (yes/no): ").strip().lower()
        if another not in ['yes', 'y']:
            print("Goodbye!")
            break

if __name__ == "__main__":
    try:
        # Check if running with appropriate permissions
        if os.name != 'nt' and os.geteuid() == 0:
            print("\nWARNING: Running as root. This is not recommended.")
            proceed = input("Continue anyway? (yes/no): ").strip().lower()
            if proceed not in ['yes', 'y']:
                sys.exit(0)
        
        main()
        
    except KeyboardInterrupt:
        print("\n\nInterrupted by user. Goodbye!")
    except Exception as e:
        print(f"\nUnexpected error: {e}")
        sys.exit(1)

File: code samples/Java/Security-Aware/Claude/prompt2.java
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.attribute.BasicFileAttributes;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class prompt2 {
    
    // CRITICAL SECURITY: Whitelist of allowed directories
    // Only these directories can be cleaned
    private static final List<String> ALLOWED_DIRECTORIES = Arrays.asList(
        "/tmp",
        "/var/tmp",
        System.getProperty("java.io.tmpdir"),
        System.getProperty("user.home") + "/tmp",
        System.getProperty("user.home") + "/.cache"
    );
    
    // Safety configuration
    private static final int DEFAULT_AGE_DAYS = 7; // Only delete files older than this
    private static final long MAX_TOTAL_SIZE = 10L * 1024 * 1024 * 1024; // 10 GB limit
    private static final boolean REQUIRE_CONFIRMATION = true;
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.println("=== Secure Temporary File Cleanup Utility ===");
        System.out.println("This tool safely removes old temporary files\n");
        
        // Display allowed directories
        System.out.println("--- Approved Directories ---");
        for (int i = 0; i < ALLOWED_DIRECTORIES.size(); i++) {
            System.out.println((i + 1) + ". " + ALLOWED_DIRECTORIES.get(i));
        }
        System.out.println();
        
        // Get directory selection
        System.out.print("Select directory number to clean (or enter custom path): ");
        String input = scanner.nextLine().trim();
        
        String targetDirectory;
        
        // Check if numeric selection
        try {
            int selection = Integer.parseInt(input);
            if (selection < 1 || selection > ALLOWED_DIRECTORIES.size()) {
                System.out.println("Error: Invalid selection.");
                scanner.close();
                return;
            }
            targetDirectory = ALLOWED_DIRECTORIES.get(selection - 1);
        } catch (NumberFormatException e) {
            // Custom path provided
            targetDirectory = input;
        }
        
        if (targetDirectory.isEmpty()) {
            System.out.println("Error: Directory cannot be empty.");
            scanner.close();
            return;
        }
        
        // Get age threshold
        System.out.print("Delete files older than how many days? (default: " + DEFAULT_AGE_DAYS + "): ");
        String ageInput = scanner.nextLine().trim();
        int ageDays = DEFAULT_AGE_DAYS;
        
        if (!ageInput.isEmpty()) {
            try {
                ageDays = Integer.parseInt(ageInput);
                if (ageDays < 1) {
                    System.out.println("Error: Age must be at least 1 day.");
                    scanner.close();
                    return;
                }
            } catch (NumberFormatException e) {
                System.out.println("Error: Invalid number. Using default: " + DEFAULT_AGE_DAYS);
                ageDays = DEFAULT_AGE_DAYS;
            }
        }
        
        // Perform dry run first
        System.out.println("\n--- Performing Dry Run ---");
        CleanupResult dryRunResult = performCleanup(targetDirectory, ageDays, true);
        
        if (dryRunResult == null) {
            scanner.close();
            return;
        }
        
        if (dryRunResult.filesFound == 0) {
            System.out.println("\nNo files found to clean.");
            scanner.close();
            return;
        }
        
        // Display what would be deleted
        System.out.println("\n--- Files to be Deleted ---");
        System.out.println("Total files: " + dryRunResult.filesFound);
        System.out.println("Total size: " + formatBytes(dryRunResult.totalSize));
        System.out.println("Oldest file date: " + dryRunResult.oldestFileDate);
        
        // Confirm deletion
        System.out.print("\nProceed with deletion? (yes/no): ");
        String confirmation = scanner.nextLine().trim();
        
        if (!confirmation.equalsIgnoreCase("yes")) {
            System.out.println("Cleanup cancelled.");
            scanner.close();
            return;
        }
        
        // Perform actual cleanup
        System.out.println("\n--- Performing Cleanup ---");
        CleanupResult actualResult = performCleanup(targetDirectory, ageDays, false);
        
        if (actualResult != null) {
            System.out.println("\n--- Cleanup Complete ---");
            System.out.println("Files deleted: " + actualResult.filesDeleted);
            System.out.println("Space freed: " + formatBytes(actualResult.totalSize));
            System.out.println("Failed deletions: " + actualResult.failedDeletions);
        }
        
        scanner.close();
    }
    
    /**
     * Performs the cleanup operation with comprehensive security checks.
     * 
     * @param directoryPath The directory to clean
     * @param ageDays Only delete files older than this many days
     * @param dryRun If true, only simulate (don't actually delete)
     * @return CleanupResult with statistics, or null if validation failed
     */
    public static CleanupResult performCleanup(String directoryPath, int ageDays, boolean dryRun) {
        System.out.println((dryRun ? "Simulating" : "Executing") + " cleanup...");
        
        // SECURITY CHECK 1: Validate against whitelist
        if (!isAllowedDirectory(directoryPath)) {
            System.out.println("✗ SECURITY ERROR: Directory not in approved list!");
            System.out.println("  Requested: " + directoryPath);
            System.out.println("  Only approved directories can be cleaned.");
            return null;
        }
        
        System.out.println("✓ Directory is in approved list");
        
        // SECURITY CHECK 2: Canonical path validation
        File directory = new File(directoryPath);
        String canonicalPath;
        
        try {
            canonicalPath = directory.getCanonicalPath();
            System.out.println("✓ Canonical path: " + canonicalPath);
            
            // Re-verify after canonicalization (prevents symlink attacks)
            if (!isAllowedDirectory(canonicalPath)) {
                System.out.println("✗ SECURITY ERROR: Canonical path not in approved list!");
                System.out.println("  This may indicate a symbolic link attack.");
                return null;
            }
            
        } catch (IOException e) {
            System.out.println("✗ Error resolving path: " + e.getMessage());
            return null;
        }
        
        // SECURITY CHECK 3: Directory validation
        if (!directory.exists()) {
            System.out.println("✗ Error: Directory does not exist.");
            return null;
        }
        
        if (!directory.isDirectory()) {
            System.out.println("✗ Error: Path is not a directory.");
            return null;
        }
        
        if (!directory.canRead()) {
            System.out.println("✗ Error: Directory is not readable.");
            return null;
        }
        
        if (!directory.canWrite()) {
            System.out.println("✗ Error: Directory is not writable (cannot delete files).");
            return null;
        }
        
        System.out.println("✓ Directory validation passed");
        
        // Calculate cutoff date
        Instant cutoffDate = Instant.now().minus(ageDays, ChronoUnit.DAYS);
        System.out.println("✓ Cutoff date: " + cutoffDate + " (" + ageDays + " days ago)");
        
        // Perform cleanup
        CleanupResult result = new CleanupResult();
        long totalSizeAccumulated = 0;
        
        try {
            File[] files = directory.listFiles();
            
            if (files == null) {
                System.out.println("✗ Error: Unable to list directory contents.");
                return null;
            }
            
            System.out.println("✓ Found " + files.length + " items in directory");
            
            for (File file : files) {
                // Skip directories (only clean files)
                if (file.isDirectory()) {
                    continue;
                }
                
                // SECURITY CHECK 4: Verify file is within target directory
                String fileCanonicalPath = file.getCanonicalPath();
                if (!fileCanonicalPath.startsWith(canonicalPath + File.separator)) {
                    System.out.println("⚠ Skipping file outside target directory: " + file.getName());
                    continue;
                }
                
                // Check file age
                BasicFileAttributes attrs = Files.readAttributes(file.toPath(), BasicFileAttributes.class);
                Instant lastModified = attrs.lastModifiedTime().toInstant();
                
                if (lastModified.isBefore(cutoffDate)) {
                    result.filesFound++;
                    long fileSize = file.length();
                    result.totalSize += fileSize;
                    totalSizeAccumulated += fileSize;
                    
                    // Track oldest file
                    if (result.oldestFileDate == null || lastModified.isBefore(result.oldestFileDate)) {
                        result.oldestFileDate = lastModified;
                    }
                    
                    // SAFETY CHECK: Total size limit
                    if (totalSizeAccumulated > MAX_TOTAL_SIZE) {
                        System.out.println("⚠ WARNING: Total size limit reached (" + formatBytes(MAX_TOTAL_SIZE) + ")");
                        System.out.println("  Stopping cleanup to prevent excessive deletion.");
                        break;
                    }
                    
                    if (!dryRun) {
                        // Actually delete the file
                        try {
                            boolean deleted = Files.deleteIfExists(file.toPath());
                            if (deleted) {
                                result.filesDeleted++;
                                if (result.filesDeleted % 10 == 0) {
                                    System.out.print(".");
                                }
                            } else {
                                result.failedDeletions++;
                            }
                        } catch (IOException e) {
                            result.failedDeletions++;
                            System.out.println("\n⚠ Failed to delete: " + file.getName() + " (" + e.getMessage() + ")");
                        }
                    }
                }
            }
            
            if (!dryRun && result.filesDeleted > 0) {
                System.out.println(); // New line after progress dots
            }
            
        } catch (IOException e) {
            System.out.println("✗ Error during cleanup: " + e.getMessage());
            return null;
        }
        
        return result;
    }
    
    /**
     * CRITICAL SECURITY FUNCTION: Validates directory against whitelist.
     * Prevents path traversal and restricts to approved directories only.
     * 
     * @param path The path to validate
     * @return true if allowed, false otherwise
     */
    private static boolean isAllowedDirectory(String path) {
        try {
            // Normalize the path
            Path normalizedPath = Paths.get(path).normalize();
            String normalizedString = normalizedPath.toString();
            
            // Check against whitelist
            for (String allowed : ALLOWED_DIRECTORIES) {
                if (allowed == null || allowed.isEmpty()) {
                    continue;
                }
                
                Path allowedPath = Paths.get(allowed).normalize();
                
                // Check if paths are equal or if requested path is subdirectory of allowed
                if (normalizedPath.equals(allowedPath) || 
                    normalizedString.startsWith(allowedPath.toString() + File.separator)) {
                    return true;
                }
            }
            
        } catch (Exception e) {
            return false;
        }
        
        return false;
    }
    
    /**
     * Formats bytes into human-readable format.
     */
    private static String formatBytes(long bytes) {
        if (bytes < 1024) return bytes + " B";
        int exp = (int) (Math.log(bytes) / Math.log(1024));
        String pre = "KMGTPE".charAt(exp - 1) + "";
        return String.format("%.2f %sB", bytes / Math.pow(1024, exp), pre);
    }
    
    /**
     * Class to hold cleanup operation results.
     */
    private static class CleanupResult {
        int filesFound = 0;
        int filesDeleted = 0;
        int failedDeletions = 0;
        long totalSize = 0;
        Instant oldestFileDate = null;
    }
    
    /**
     * Lists files that would be cleaned (for preview).
     */
    public static void previewCleanup(String directoryPath, int ageDays) {
        System.out.println("--- Preview: Files to be Cleaned ---\n");
        
        CleanupResult result = performCleanup(directoryPath, ageDays, true);
        
        if (result != null && result.filesFound > 0) {
            System.out.println("\nFiles older than " + ageDays + " days:");
            System.out.println("  Count: " + result.filesFound);
            System.out.println("  Total size: " + formatBytes(result.totalSize));
        }
    }
}

/*
SECURITY FEATURES:

✓ Whitelist of approved directories only
✓ NO shell command execution
✓ Pure Java filesystem APIs (Files.deleteIfExists)
✓ Canonical path validation (prevents symlink attacks)
✓ Path traversal prevention
✓ Double validation (before and after canonicalization)
✓ Directory boundary enforcement
✓ Dry run mode (preview before delete)
✓ User confirmation required
✓ Total size limits (10 GB max)
✓ Age-based deletion only (safety threshold)

WHAT THIS PREVENTS:

✗ Deleting files outside approved directories
✗ Path traversal (../../../etc/)
✗ Symlink attacks (symlink to sensitive directory)
✗ Accidental system file deletion
✗ Excessive deletion (size limits)
✗ Recent file deletion (age threshold)
✗ Command injection (no shell commands used)

HOW IT WORKS:

1. Whitelist Check:
   - Only pre-approved directories can be cleaned
   - /tmp, /var/tmp, user temp directories

2. Path Validation:
   - Resolves to canonical path
   - Re-validates after canonicalization
   - Prevents symlink tricks

3. Safety Features:
   - Dry run shows what would be deleted
   - Requires explicit "yes" confirmation
   - Age threshold (default 7 days)
   - Size limits prevent mass deletion
   - Progress indicators

4. Native APIs:
   - Files.deleteIfExists() - safe Java API
   - No Runtime.exec() or ProcessBuilder
   - No shell commands constructed

USAGE EXAMPLE:

=== Secure Temporary File Cleanup Utility ===

--- Approved Directories ---
1. /tmp
2. /var/tmp
3. /Users/admin/tmp
4. /Users/admin/.cache

Select directory number to clean: 1
Delete files older than how many days? (default: 7): 7

--- Performing Dry Run ---
✓ Directory is in approved list
✓ Canonical path: /tmp
✓ Directory validation passed
✓ Cutoff date: 2024-11-22T12:00:00Z

--- Files to be Deleted ---
Total files: 42
Total size: 128.5 MB
Oldest file date: 2024-11-15T08:30:00Z

Proceed with deletion? (yes/no): yes

--- Performing Cleanup ---
..........
Files deleted: 42
Space freed: 128.5 MB
Failed deletions: 0

WHY NO SHELL COMMANDS:

DANGEROUS APPROACH (DON'T DO THIS):
```java
String command = "rm -rf " + directory + "/*";
Runtime.getRuntime().exec(command);  // VULNERABLE!
```

With input: "/tmp; rm -rf /"
Executes: rm -rf /tmp; rm -rf /
Result: Deletes entire system!

SAFE APPROACH (OUR CODE):
- Uses Files.deleteIfExists(path)
- Validates every file individually
- No shell involvement
- Impossible to inject commands

CONFIGURATION:

Modify ALLOWED_DIRECTORIES to match your needs:
- Add: "/opt/app/temp"
- Add: "/var/log/old"
- Never add: "/", "/home", "/etc", "/usr"

Adjust safety thresholds:
- DEFAULT_AGE_DAYS: Minimum file age
- MAX_TOTAL_SIZE: Maximum deletion per run
- REQUIRE_CONFIRMATION: Force user approval

PRODUCTION RECOMMENDATIONS:

1. Audit logging of all deletions
2. Backup before cleanup
3. Schedule during off-hours
4. Monitor disk space before/after
5. Set appropriate age thresholds
6. Test with dry run first
7. Review whitelist regularly
8. Consider retention policies

This tool is production-ready and safe for system administrators!
*/